### PR TITLE
[ARCTIC-1783][lookup] The primary key serialization exception occurs when the primary key definition is different from the schema fields order

### PR DESCRIPTION
**Pull request by YesOrNo828**
## Why are the changes needed?

fixed: #1783 

## Brief change log

  - *Adjust the order of the primary keys and join keys when creating unique and join keys mapping index*
  - *The Flink 1.15 and 1.14 versions are affected.*
  - *Avoid NPE if the initialization process fails*
  - *Add some tests*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java [hunk 1]
```diff
@@ -71,6 +71,7 @@ public abstract class RocksDBCacheState<V> {
   protected BinaryRowDataSerializerWrapper valueSerializer;
   private ExecutorService writeRocksDBService;
   private final AtomicBoolean initialized = new AtomicBoolean(false);
+  private final AtomicBoolean running = new AtomicBoolean(true);
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java:74]: Is the variable name `closed` better here?
I do not think the class is runnable.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/KVTableFactory.java [hunk 2]
```diff
@@ -46,6 +48,12 @@ public KVTable<RowData> create(
       Predicate<RowData> rowDataPredicate) {
     Set<String> joinKeySet = new HashSet<>(joinKeys);
     Set<String> primaryKeySet = new HashSet<>(primaryKeys);
+    // keep the primary keys order with projected schema fields.
+    primaryKeys = projectSchema.asStruct().fields().stream()
```
Comment (zhoujinsong) [inline flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/KVTableFactory.java:52]: Should the order of `joinKeys` be adjust too here?

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java (hunk 1 above)
Comment (YesOrNo828) [reply flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/RocksDBCacheState.java]: Make sense to me.

File: flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/KVTableFactory.java (hunk 2 above)
Comment (YesOrNo828) [reply flink/v1.14/flink/src/main/java/com/netease/arctic/flink/lookup/KVTableFactory.java:52]: I add a test about this case.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**